doc/lms_code_walk.md

```diff
@@ -23,8 +23,165 @@ into intermediate representations (IR). Instead, the IR is generated via executi
 [LMS Simple Frontend](main/scala/lms/core/frontend.md)
 
+The iconic typed (`Rep[T]`) frondend is defined in this file `lms-clean/src/main/scala/lms/core/stub.scala`:
 [LMS Frontend](main/scala/lms/core/stub.md)
+
+The basic codegen support is in the file `lms-clean/src/main/scala/lms/core/codegen.scala`
+and the folder `lms-clean/src/main/scala/lms/core/codegen/*`.
+TODO(feiw): add a markdown file for codegen.
+
+
+### Extension
+
+Besides the basic LMS support mensioned above, LMS can be extended to support the following:
+1. Collections (Array, List, Tuple, ...)
+2. ThirdParty library (MPI, NCCL, ...)
+3. Other Types (fp16, tensor, ...)
+
+The extensions normally include implementations of:
+1. Extension of FrontEnd (traits of construction of new ops).
+2. Extension of CodeGen (how new ops are generated to the target language).
+3. Extension of Graph Transformation (how new ops are transformed).
+
+### Collections
+
+Several examples of collections are implemented here (`lms-clean/src/main/scala/lms/collection`). Taking the ListOps.scala as example, it has four traits in total.
+
+The `trait ListOps` is the frontend trait of `Rep[List]`. It contains an `object List` for the construction of `Rep[List]`-typed values, and an `implicit class ListOps` for the methods that are available for `Rep[List]`-typed values.
+
+The `trait ListOpsOpt` is the optimized frontend trait of `Rep[List]`, where several
+construction-time optimization are introduced.
+
+The `trait ScalaCodeGen_List` is the trait for generating scala code from the nodes
+created by `trait ListOps` and `trait ListOpsOpt`. The most important function in this
+trait is `override def shallow` to generate code for each `Rep[List]` node.
+
+The `trait CppCodeGen_List` is the trait for generating C++ code from the nodes
+created by `trait ListOps` and `trait ListOpsOpt`. Similarly, the most important
+function in this trait is `override def shallow` that generates code for each `Rep[List]` node.
+
+### ThirdParty Library
+
+Several examples of third party libraries are implemented here (`lms-clean/src/main/scala/lms/thirdparty`). The key functionality is provide by `lms-clean/src/main/scala/lms/thirdparty/c_lib_utils.scala`.
+
+In the `c_lib_utils.scala` file, we provide the following functionalities:
+1. Create a Macro by `def cmacro` function.
+2. Create a C Struct by `def newStruct` function.
+3. Create a call to library function by `def libFunction` function.
+
+We only create three types of nodes for thirdparty libraries.
+1. The `def cmacro` function creates nodes with `op = "cmacro"`.
+2. The `def newStruct` function creates nodes with `op = "lib-struct"`.
+3. The `def libFunction` function creates nodes with `op = "lib-Function"`.
+
+The codegen support those nodes are in this same file.
+To mix in traits for the frontend support, just extend the `trait CLibs`.
+To mix in traits for the codegen support, just extend the `trait CCodeGenLibs`.
```

Comment:
mix in -> mixin?

Reply:
I guess `mix in` is verb and `mixin` is noun?